src/charm.py

```diff
@@ -1,199 +1,222 @@
 
 logger = logging.getLogger(__name__)
 
 
 class NatsStartedEvent(EventBase):
+    """NATS Started Event."""
+
     pass
 
 
 class NatsCharmEvents(CharmEvents):
+    """NATS Custom Charm Events."""
+
     nats_started = EventSource(NatsStartedEvent)
 
 
+SNAP_COMMON_PATH = Path("/var/snap/nats/common")
+SERVER_PATH = SNAP_COMMON_PATH / "server"
+
+
 class NatsCharm(CharmBase):
-    on = NatsCharmEvents()
+    """Charmed Operator to deploy NATS - a distributed message bus for services."""
+
+    on: CharmEvents = NatsCharmEvents()
     state = StoredState()
 
-    NATS_SERVICE = 'snap.nats.server.service'
-    SNAP_COMMON_PATH = Path('/var/snap/nats/common')
-    SERVER_PATH = SNAP_COMMON_PATH / 'server'
-    NATS_SERVER_CONFIG_PATH = SERVER_PATH / 'nats.cfg'
-    AUTH_TOKEN_PATH = SERVER_PATH / 'auth_secret'
+    NATS_SERVICE = "snap.nats.server.service"
+    NATS_SERVER_CONFIG_PATH = SERVER_PATH / "nats.cfg"
+    AUTH_TOKEN_PATH = SERVER_PATH / "auth_secret"
     AUTH_TOKEN_LENGTH = 64
-    TLS_KEY_PATH = SERVER_PATH / 'key.pem'
-    TLS_CERT_PATH = SERVER_PATH / 'cert.pem'
-    TLS_CA_CERT_PATH = SERVER_PATH / 'ca.pem'
-
-    def __init__(self, framework, key):
-        super().__init__(framework, key)
-
-        for event in (self.on.install,
-                      self.on.start,
-                      self.on.upgrade_charm,
-                      self.on.config_changed,
-                      self.on.cluster_relation_changed,
-                      self.on.client_relation_joined):
-            self.framework.observe(event, self)
-
-        listen_on_all_addresses = self.model.config['listen-on-all-addresses']
-        self.cluster = NatsCluster(self, 'cluster', listen_on_all_addresses)
-        self.client = NatsClient(self, 'client', listen_on_all_addresses, self.model.config['client-port'])
-        self.state.set_default(is_started=False, auth_token=self.get_auth_token(self.AUTH_TOKEN_LENGTH),
-                               use_tls=None, use_tls_ca=None, nats_config_hash=None,
-                               client_port=None)
-
-        self.ca_client = CAClient(self, 'ca-client')
-        self.framework.observe(self.ca_client.on.tls_config_ready, self)
-        self.framework.observe(self.ca_client.on.ca_available, self)
-
-        self.nrpe_client = NRPEClient(self, 'nrpe-external-master')
-        self.framework.observe(self.nrpe_client.on.nrpe_available, self)
-
-    def on_install(self, event):
+    TLS_KEY_PATH = SERVER_PATH / "key.pem"
+    TLS_CERT_PATH = SERVER_PATH / "cert.pem"
+    TLS_CA_CERT_PATH = SERVER_PATH / "ca.pem"
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self.state.set_default(
+            is_started=False,
+            auth_token=NatsCharm.get_auth_token(self.AUTH_TOKEN_LENGTH),
+            use_tls=None,
+            use_tls_ca=None,
+            nats_config_hash=None,
+            client_port=None,
+        )
```

Comment:
In general you want to avoid `StoredState`. It's used incorrectly very frequently - and caching things like `is_started` is a very bad idea. If you need persistent data I'd recommend using the peer relation.